File: 3.Stacks and Queues/3.StackUsingQueue.js
class StackUsingQueue {
    constructor() {
        this.mainQueue = [];
        this.tempQueue = [];
    }

    push(value) {
        while (this.mainQueue.length !== 0) {
            // Move all elements from mainQueue to tempQueue using shift operation
            this.tempQueue.push(this.mainQueue.shift());
        }
    
        this.mainQueue.push(value);
    
        while (this.tempQueue.length !== 0) {
            // Move all elements from tempQueue to mainQueue using shift operation
            this.mainQueue.push(this.tempQueue.shift());
        }
    }

    pop() {
        if (this.mainQueue.length === 0) {
            return -1;
        }
        return this.mainQueue.shift();
    }

    top() {
        if (this.mainQueue.length === 0) {
            return -1;
        }
        return this.mainQueue[0];
    }

    empty() {
        return this.mainQueue.length === 0;
    }
}

File: 3.Stacks and Queues/3.StackUsingQueue.py
class StackUsingQueue:
    def __init__(self):
        self.main_queue = []
        self.temp_queue = []

    def push(self, value):
        while len(self.main_queue) != 0:
            # move all elements from main_queue to temp_queue using shift operation (pop(0))
            self.temp_queue.append(self.main_queue.pop(0))
        self.main_queue.append(value)
        while len(self.temp_queue) != 0:
            # move all elements from temp_queue to main_queue using shift operation (pop(0))
            self.main_queue.append(self.temp_queue.pop(0))

    def pop(self):
        return self.main_queue.pop(0)
    
    def peek(self):
        return self.main_queue[0]
    
    def is_empty(self):
        return len(self.main_queue) == 0
